File: 프로그래머스/2/169198. 당구 연습/당구 연습.kt
class Solution {
    fun solution(m: Int, n: Int, startX: Int, startY: Int, balls: Array<IntArray>): IntArray {
        val answer = IntArray(balls.size)

        for (i in balls.indices) {
            val (a, b) = balls[i]
            var minDist = Int.MAX_VALUE

            // 1️⃣ 위쪽 벽 반사 (y → 2n - b)
            if (!(startX == a && startY < b)) {
                val y = 2 * n - b
                val dist = distSq(startX, startY, a, y)
                minDist = minOf(minDist, dist)
            }

            // 2️⃣ 아래쪽 벽 반사 (y → -b)
            if (!(startX == a && startY > b)) {
                val y = -b
                val dist = distSq(startX, startY, a, y)
                minDist = minOf(minDist, dist)
            }

            // 3️⃣ 왼쪽 벽 반사 (x → -a)
            if (!(startY == b && startX > a)) {
                val x = -a
                val dist = distSq(startX, startY, x, b)
                minDist = minOf(minDist, dist)
            }

            // 4️⃣ 오른쪽 벽 반사 (x → 2m - a)
            if (!(startY == b && startX < a)) {
                val x = 2 * m - a
                val dist = distSq(startX, startY, x, b)
                minDist = minOf(minDist, dist)
            }

            answer[i] = minDist
        }

        return answer
    }

    // 거리의 제곱 계산 함수
    private fun distSq(x1: Int, y1: Int, x2: Int, y2: Int): Int {
        val dx = x1 - x2
        val dy = y1 - y2
        return dx * dx + dy * dy
    }
}

File: 프로그래머스/2/169198. 당구 연습/당구 연습.py
def solution(m, n, startX, startY, balls):
    answer = []
    
    for a, b in balls:
        min_dist = float('inf')

        # 1️⃣ 위쪽 벽 반사
        if not (startX == a and startY < b):  # 같은 세로선 위에서 위로 직접 맞는 경우 제외
            y = 2 * n - b
            dist = (startX - a) ** 2 + (startY - y) ** 2
            min_dist = min(min_dist, dist)

        # 2️⃣ 아래쪽 벽 반사
        if not (startX == a and startY > b):  # 아래로 직접 맞는 경우 제외
            y = -b
            dist = (startX - a) ** 2 + (startY - y) ** 2
            min_dist = min(min_dist, dist)

        # 3️⃣ 왼쪽 벽 반사
        if not (startY == b and startX > a):  # 왼쪽으로 직접 맞는 경우 제외
            x = -a
            dist = (startX - x) ** 2 + (startY - b) ** 2
            min_dist = min(min_dist, dist)

        # 4️⃣ 오른쪽 벽 반사
        if not (startY == b and startX < a):  # 오른쪽으로 직접 맞는 경우 제외
            x = 2 * m - a
            dist = (startX - x) ** 2 + (startY - b) ** 2
            min_dist = min(min_dist, dist)

        answer.append(min_dist)

    return answer
